src/sqlfluff/dialects/dialect_spark3.py

```diff
@@ -1086,6 +1085,143 @@ class SelectStatementSegment(BaseSegment):
     )
 
 
+@spark3_dialect.segment(replace=True)
+class GroupByClauseSegment(BaseSegment):
+    """Enhance `GROUP BY` clause like in `SELECT` for 'CUBE' and 'ROLLUP`.
+
+    https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-groupby.html
+    """
+
+    type = "group_by_clause"
+
+    match_grammar = StartsWith(
+        Sequence("GROUP", "BY"),
+        terminator=OneOf("ORDER", "LIMIT", "HAVING", "WINDOW"),
+        enforce_whitespace_preceding_terminator=True,
+    )
+
+    parse_grammar = Sequence(
+        "GROUP",
+        "BY",
+        Indent,
+        Delimited(
+            OneOf(
+                Ref("ColumnReferenceSegment"),
+                # Can `GROUP BY 1`
+                Ref("NumericLiteralSegment"),
+                # Can `GROUP BY coalesce(col, 1)`
+                Ref("ExpressionSegment"),
+                Ref("CubeRollupClauseSegment"),
+                Ref("GroupingSetsClauseSegment"),
+            ),
+            terminator=OneOf("ORDER", "LIMIT", "HAVING", "WINDOW"),
+        ),
+        # TODO: New Rule
+        #  Warn if CubeRollupClauseSegment and
+        #  WithCubeRollupClauseSegment used in same query
+        Ref("WithCubeRollupClauseSegment", optional=True),
+        Dedent,
+    )
+
+
+@spark3_dialect.segment()
+class WithCubeRollupClauseSegment(BaseSegment):
+    """A `[WITH CUBE | WITH ROLLUP]` clause after the `GROUP BY` clause.
+
+    https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-groupby.html
+    """
+
+    type = "with_cube_rollup_clause"
+
+    match_grammar = StartsWith(
+        Sequence(
+            "WITH",
+            OneOf("CUBE", "ROLLUP"),
+        ),
+        terminator=OneOf(
```

Comment:
Why do you need this? It’s a simple enough statement so shouldn’t be confused with the next statement (which is the usual reason for adding a terminator clause).

So can this `match_clause` and `parse_clause`s be merged into one simple `match_clause`?

Reply:
yes, I'll make that update. They can be merged